### PR TITLE
Scaling done

### PR DESCRIPTION
**Pull request by borcuttjahns**

### CODE REVIEW

**Comment by borcuttjahns** [comment]
Let me know what you think - there's definitely some mild wonkiness (I'd guess from scaling IL2 and IL15 together?) but it doesn't look bad, particularly for T-regs. If you like this we can move ahead with the muteins.

File: ckine/figures/figureCommon.py [hunk 1]
```diff
@@ -437,3 +438,53 @@ def hill_equation(x, x0, solution=0):
...
+        output2[(iterator * 4): (iteratorEnd * 4)] = expSTAT2[(iterator * 4): (iteratorEnd * 4)] * lsqScale.x[1] + lsqScale.x[0]
+        output15[(iterator * 4): (iteratorEnd * 4)] = expSTAT15[(iterator * 4): (iteratorEnd * 4)] * lsqScale.x[1] + lsqScale.x[0]
+        iterator = copy.copy(iteratorEnd)
+
+    return output2, output15
+
+
+def scaleResids(x, preds, exps):
+    """Residual function for linear scaling"""
+    return np.linalg.norm((exps * x[1] + x[0]) - preds)
```
Comment (aarmey) [inline ckine/figures/figureCommon.py:490]: Use this which will ensure you're getting the optimum efficiently:

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.linregress.html

If the data is in matrix form you can use `ndarray.ravel()` to flatten it.

File: ckine/figures/figureCommon.py [hunk 2]
```diff
@@ -437,3 +438,53 @@ def hill_equation(x, x0, solution=0):
...
+        subExpSTAT15 = np.reshape(subExpSTAT15, (subExpSTAT15.shape[0], subExpSTAT15.shape[1], subExpSTAT15.shape[2], 1))
+        subExpSTAT15 = np.tile(subExpSTAT15, (1, 1, 1, subPredSTAT15.shape[3]))
+
+        expSTAT = np.vstack((subExpSTAT2, subExpSTAT15))
+        predSTAT = np.vstack((subPredSTAT2, subPredSTAT2))
+
+        lsqScale = least_squares(scaleResids, x0, args=(predSTAT, expSTAT))
+        output2[(iterator * 4): (iteratorEnd * 4)] = expSTAT2[(iterator * 4): (iteratorEnd * 4)] * lsqScale.x[1] + lsqScale.x[0]
+        output15[(iterator * 4): (iteratorEnd * 4)] = expSTAT15[(iterator * 4): (iteratorEnd * 4)] * lsqScale.x[1] + lsqScale.x[0]
+        iterator = copy.copy(iteratorEnd)
```
Comment (aarmey) [inline ckine/figures/figureCommon.py:483]: If this is a standard float you shouldn't need copy... if it's a numpy array you can use `ndarray.copy()`.

File: ckine/figures/figureCommon.py [hunk 3]
```diff
@@ -437,3 +438,53 @@ def hill_equation(x, x0, solution=0):
...
+        subPredSTAT2 = predSTAT2[(iterator): (iteratorEnd), :, :, :]
+        subPredSTAT15 = predSTAT15[(iterator): (iteratorEnd), :, :, :]
+
+        subPredSTAT2 = np.swapaxes(subPredSTAT2, 3, 1)
+        subPredSTAT2 = np.swapaxes(subPredSTAT2, 3, 2)
+        subPredSTAT15 = np.swapaxes(subPredSTAT15, 3, 1)
+        subPredSTAT15 = np.swapaxes(subPredSTAT15, 3, 2)
+
+        if rep2:
+            subExpSTAT2 = np.nan_to_num(subExpSTAT2)
```
Comment (aarmey) [inline ckine/figures/figureCommon.py:464]: You're replacing nan's with 0's? Isn't that going to affect the results?

**Review by aarmey — CHANGES_REQUESTED** [review]
A couple small edits.

File: ckine/figures/figureCommon.py (hunk 3 above)
Comment (borcuttjahns) [reply ckine/figures/figureCommon.py]: Was an oversight - replaced with 0 for 2 sets of 2 data points in the IL15 experiments as it was low concentrations. I've replaced this by copying the third smallest concentration into those slots.

File: ckine/figures/figureCommon.py (hunk 1 above)
Comment (borcuttjahns) [reply ckine/figures/figureCommon.py]: Just took care of this, but for some reason, while the slope matches my previous method, the intercept changed significantly, which makes the fit look significantly worse at lower concentrations. I'm guessing the fit is better though in terms of actual mean squared error.

File: ckine/figures/figureCommon.py (hunk 3 above)
Comment (aarmey) [reply ckine/figures/figureCommon.py]: I think you're better off just removing values for the conversion calculation when one is nan. You can do ravel, then isnan to identify which indices are nan, then remove those indices from x and y.

Comment (borcuttjahns) [reply ckine/figures/figureCommon.py]: Gotcha - taken care of.

**Comment by aarmey** [comment]
Excellent—will merge this now. Indeed, it looks a little wonky, but I think that's mostly because this experiment saw a very large reduction in pSTAT5 at 4 hrs. Setup the muteins now and let's see what those look like. Depending on those results, we can consider discarding the 4 hr timepoints.

**Pull request merged**